geth/api/backend.go

```diff
@@ -77,120 +78,97 @@ func (m *StatusBackend) IsNodeRunning() bool {
 }
 
 // StartNode start Status node, fails if node is already started
-func (m *StatusBackend) StartNode(config *params.NodeConfig) (<-chan struct{}, error) {
-	m.Lock()
-	defer m.Unlock()
-
-	if m.nodeReady != nil {
-		return nil, node.ErrNodeExists
-	}
+func (m *StatusBackend) StartNode(config *params.NodeConfig) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.startNode(config)
+}
 
-	nodeStarted, err := m.nodeManager.StartNode(config)
+func (m *StatusBackend) startNode(config *params.NodeConfig) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("node crashed on start: %v", err)
+		}
+	}()
+	err = m.nodeManager.StartNode(config)
 	if err != nil {
-		return nil, err
+		switch err.(type) {
+		case node.RPCClientError:
+			err = node.ErrRPCClient
+		case node.EthNodeError:
+			err = fmt.Errorf("%v: %v", node.ErrNodeStartFailure, err)
+		}
+		signal.Send(signal.Envelope{
+			Type: signal.EventNodeCrashed,
+			Event: signal.NodeCrashEvent{
+				Error: err,
+			},
+		})
+		return err
 	}
-
-	m.nodeReady = make(chan struct{}, 1)
-	go m.onNodeStart(nodeStarted, m.nodeReady) // waits on nodeStarted, writes to backendReady
-
-	return m.nodeReady, err
+	// not sure if we should keep both, onNodeStart is so simple, that it doesn't
+	signal.Send(signal.Envelope{Type: signal.EventNodeStarted})
+	m.onNodeStart()
+	signal.Send(signal.Envelope{Type: signal.EventNodeReady})
+	return nil
 }
 
 // onNodeStart does everything required to prepare backend
-func (m *StatusBackend) onNodeStart(nodeStarted <-chan struct{}, backendReady chan struct{}) {
-	<-nodeStarted
-
+func (m *StatusBackend) onNodeStart() {
 	// tx queue manager should be started after node is started, it depends
 	// on rpc client being created
 	m.txQueueManager.Start()
-
 	if err := m.registerHandlers(); err != nil {
 		log.Error("Handler registration failed", "err", err)
 	}
-
 	if err := m.accountManager.ReSelectAccount(); err != nil {
 		log.Error("Reselect account failed", "err", err)
 	}
 	log.Info("Account reselected")
-
-	close(backendReady)
-	signal.Send(signal.Envelope{
-		Type:  signal.EventNodeReady,
-		Event: struct{}{},
-	})
 }
 
 // StopNode stop Status node. Stopped node cannot be resumed.
-func (m *StatusBackend) StopNode() (<-chan struct{}, error) {
-	m.Lock()
-	defer m.Unlock()
+func (m *StatusBackend) StopNode() error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.stopNode()
+}
 
-	if m.nodeReady == nil {
-		return nil, node.ErrNoRunningNode
+func (m *StatusBackend) stopNode() error {
+	if !m.IsNodeRunning() {
```

Comment:
why isn't it a responsibility of `NodeManager`?
nothing in this method is specific for the backend struct...

Reply:
i believe this is just a shortcut

Reply:
sure, what I meant is that his code doesn't seems to belong to `StatusBackend`, instead it seems to belong to `NodeManager`.
So we can move down this `if` and then
we will be able to nuke the `stopNode()` method from `StatusBackend()`.

Reply:
we won't be able to nuke it completely because this is a lock-free version that is used by other methods, but i think this condition can be dropped

Reply:
ah, I didn't notice `m.txQueueManager.Stop()` still being there.
then it makes sense to keep it here.